src/translator/quality_estimator.h

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "annotation.h"
+#include "response.h"
+#include "translator/history.h"
+
+namespace marian::bergamot {
+
+class QualityEstimator {
+ public:
+  /// Computes quality-scores using values from Histories and subword tokens which comes from Response
+  ///
+  ///
+  /// @param [in] histories: Histories obtained from translating a blob of source-text
+  /// @param [in] response: Partially constructed response, holding tokenization info
+  /// for source and target. The quality-scores for each sentence obtained from source-text blob
+  /// are written out as SentenceQualityEstimate into response.
+  virtual void computeQualityScores(const Histories &histories, Response &response) const = 0;
+};
+
+using SubwordRange = ByteRange;
+
+/// Unsupervised Quality Estimator model. It uses the translator model's log probabilities (log probs) as a proxy for
+/// quality scores. Then, for a given word, its quality score is computed by taking the mean of the log probs of the
+/// tokens that make it up. The sentence score is the mean of all word's log probs.
+class UnsupervisedQualityEstimator : public QualityEstimator {
+ public:
+  void computeQualityScores(const Histories &histories, Response &response) const override;
+
+ private:
+  Response::SentenceQualityScore computeSentenceScores(const std::vector<float> &logProbs, const AnnotatedText &target,
+                                                       const size_t sentenceIdx) const;
+};
+
+// ASCII and Unicode text files never start with the following 64 bits
+// It serves as a signature for quality estimator binary files
+constexpr std::uint64_t BINARY_QE_MODEL_MAGIC = 0x78cc336f1d54b180;
```

Comment:
Will this be the case for all binaries of different QE model types? If each has it's own and is used as a switch at load based on a header this would go into logistic regressor, would it not?

Reply:
btw, how do you know that this magic no. is not used by any other file?

Reply:
We never know that a magic number isn't used by another file, we just hope it's improbable.